src/jit/codegenxarch.cpp

```diff
@@ -3269,25 +3269,19 @@ void CodeGen::genCodeForCpBlkRepMovs(GenTreeBlk* cpBlkNode)
     GenTree* source  = cpBlkNode->Data();
     GenTree* srcAddr = nullptr;
 
-#ifdef DEBUG
     assert(dstAddr->isUsedFromReg());
     assert(source->isContained());
 
 #ifdef _TARGET_X86_
-    if (size == 0)
+    if (!cpBlkNode->OperIs(GT_STORE_DYN_BLK))
     {
-        noway_assert(cpBlkNode->OperGet() == GT_STORE_DYN_BLK);
+        assert((size == 0) || (size > CPBLK_UNROLL_LIMIT));
```

Comment:
This and the similar `assert` below are pointless. It's not like the `rep movsb` is valid only for certain sizes. FWIW one of my WIP PR removes theses.

Reply:
Also, the `size == 0` should be handled in lowering by removing the store.

Reply:
> Also, the size == 0 should be handled in lowering by removing the store.

That would not be correct without replacing it by a nullcheck of both addresses.

Reply:
Do you mean that a 0 sized `cpblk` with null src/dst is required to throw an exception? If that's really the case then the current code does not achieve that.

Reply:
I believe that the current code does indeed throw an exception. I wasn't able to discern the behavior from the spec, but I determined empirically that both `rep movs` and the helper method will throw for a null argument, even if the size is 0 (see the test I added). It's still possible I've missed something, so I'm happy to be corrected.

Reply:
> but I determined empirically that both rep movs and the helper method will throw for a null argument

I'm not sure how you did that but I've determined exactly the opposite:
* `REP MOVS` with `ecx` = 0 has no effect, the manual describes the operation of `REP` using `WHILE CountReg ≠ 0`. If you doubt the manual then it's easy to test this by throwing some inline assembly code in a VC++ x86 project:
```C++
	__asm xor esi, esi
	__asm xor edi, edi
	__asm xor ecx, ecx
	__asm rep movsb
```
* JIT_MemCpy does not have any explicit null checks: https://github.com/dotnet/coreclr/blob/030a3ea9b8dbeae89c90d34441d4d9a1cf4a7de6/src/vm/amd64/CrtHelpers.asm#L151

> even if the size is 0 (see the test I added)

That's because the test is misleading in this regard. The exception is thrown by `ref source[0]` and not by `cpblk`.